test/token/ERC20WithPermit.test.js

```diff
@@ -0,0 +1,1102 @@
+          initialSupply
+        )
+      })
+    })
+  })
+
+  describe("transfer", () => {
+    context("when the recipient is not the zero address", () => {
+      context("when the sender does not have enough balance", () => {
+        const amount = initialSupply.add(1)
+
+        it("should revert", async () => {
+          await expect(
+            token.connect(initialHolder).transfer(recipient.address, amount)
+          ).to.be.revertedWith("Transfer amount exceeds balance")
+        })
+      })
+
+      context("when the sender transfers all balance", () => {
+        const amount = initialSupply
+
+        it("should transfer the requested amount", async () => {
+          await token.connect(initialHolder).transfer(recipient.address, amount)
+
+          expect(await token.balanceOf(initialHolder.address)).to.equal(0)
+
+          expect(await token.balanceOf(recipient.address)).to.equal(amount)
+        })
+
+        it("should emit a transfer event", async () => {
+          const tx = await token
+            .connect(initialHolder)
+            .transfer(recipient.address, amount)
+
+          await expect(tx)
+            .to.emit(token, "Transfer")
+            .withArgs(initialHolder.address, recipient.address, amount)
+        })
+      })
+
+      context("when the sender transfers zero tokens", () => {
+        const amount = ethers.BigNumber.from(0)
+
+        it("should transfer the requested amount", async () => {
+          await token.connect(initialHolder).transfer(recipient.address, amount)
+
+          expect(await token.balanceOf(initialHolder.address)).to.equal(
+            initialSupply
+          )
+
+          expect(await token.balanceOf(recipient.address)).to.equal(0)
+        })
+
+        it("should emit a transfer event", async () => {
+          const tx = await token
+            .connect(initialHolder)
+            .transfer(recipient.address, amount)
+
+          await expect(tx)
+            .to.emit(token, "Transfer")
+            .withArgs(initialHolder.address, recipient.address, amount)
+        })
+      })
+    })
+
+    context("when the recipient is the zero address", () => {
+      it("should revert", async () => {
+        await expect(
+          token.connect(initialHolder).transfer(ZERO_ADDRESS, initialSupply)
+        ).to.be.revertedWith("Transfer to the zero address")
+      })
+    })
+  })
+
+  describe("transferFrom", () => {
+    context("when the token owner is not the zero address", () => {
+      context("when the recipient is not the zero address", () => {
+        context("when the spender has enough approved balance", () => {
+          const allowance = initialSupply
+          beforeEach(async function () {
+            await token
+              .connect(initialHolder)
+              .approve(anotherAccount.address, allowance)
+          })
+
+          context("when the token owner has enough balance", () => {
+            const amount = initialSupply
```

Comment:
Let's use amounts different than `initialSupply`, so we can compare that exactly the amount got transfered and tokens remained on sender account is a specific decreased amount, but not `0`.